src/Microsoft.AspNetCore.Authentication/AuthenticationMiddleware.cs

```diff
@@ -2,90 +2,64 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Text.Encodings.Web;
 using System.Threading.Tasks;
-using Microsoft.AspNetCore.Builder;
 using Microsoft.AspNetCore.Http;
-using Microsoft.Extensions.Logging;
-using Microsoft.Extensions.Options;
+using Microsoft.Extensions.DependencyInjection;
 
 namespace Microsoft.AspNetCore.Authentication
 {
-    public abstract class AuthenticationMiddleware<TOptions> where TOptions : AuthenticationOptions, new()
+    public class AuthenticationMiddleware
     {
         private readonly RequestDelegate _next;
 
-        protected AuthenticationMiddleware(
-            RequestDelegate next,
-            IOptions<TOptions> options,
-            ILoggerFactory loggerFactory,
-            UrlEncoder encoder)
+        public AuthenticationMiddleware(RequestDelegate next, IAuthenticationSchemeProvider schemes)
         {
-            if (next == null)
-            {
-                throw new ArgumentNullException(nameof(next));
-            }
-
-            if (options == null)
-            {
-                throw new ArgumentNullException(nameof(options));
-            }
-
-            if (loggerFactory == null)
-            {
-                throw new ArgumentNullException(nameof(loggerFactory));
-            }
-
-            if (encoder == null)
-            {
-                throw new ArgumentNullException(nameof(encoder));
-            }
-
-            Options = options.Value;
-            Logger = loggerFactory.CreateLogger(this.GetType().FullName);
-            UrlEncoder = encoder;
-
-            if (string.IsNullOrEmpty(Options.ClaimsIssuer))
-            {
-                // Default to something reasonable
-                Options.ClaimsIssuer = Options.AuthenticationScheme;
-            }
-
-            _next = next;
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+            Schemes = schemes ?? throw new ArgumentNullException(nameof(schemes));
         }
 
-        public string AuthenticationScheme { get; set; }
-
-        public TOptions Options { get; set; }
-
-        public ILogger Logger { get; set; }
-
-        public UrlEncoder UrlEncoder { get; set; }
+        public IAuthenticationSchemeProvider Schemes { get; set; }
 
         public async Task Invoke(HttpContext context)
         {
-            var handler = CreateHandler();
-            await handler.InitializeAsync(Options, context, Logger, UrlEncoder);
+            var oldFeature = context.Features.Get<IAuthenticationFeature>();
```

Comment:
Is it important to null out features after we are done?  Or can I nuke the try/finally entirely?